python/ray/llm/_internal/batch/stages/vllm_engine_stage.py

```diff
@@ -0,0 +1,530 @@
+        # in a separate process, the benefit of decoupling them in the Processor
+        # may be limited.
+        assert request.prompt
+        multi_modal_data = {"image": request.images} if request.images else None
+        llm_prompt = self.vllm.inputs.data.TextPrompt(
+            prompt=request.prompt, multi_modal_data=multi_modal_data
+        )
+
+        # Send the request to the LLM engine.
+        stream = self.engine.generate(
+            request_id=str(request.request_id),
+            prompt=llm_prompt,
+            sampling_params=request.params,
+        )
+
+        # Consume the stream until the request is finished.
+        async for request_output in stream:
+            if request_output.finished:
+                # Bypass the original full prompt.
+                request_output.prompt = request.prompt
+                return request_output
+
+        raise RuntimeError("Should not reach here")
+
+    def shutdown(self):
+        """Shutdown the vLLM v1 engine. This kills child processes forked
+        by the vLLM engine. If not called, the child processes will be
+        orphaned and will not be killed when the parent process exits,
+        and they won't be able to be tracked by Ray anymore.
+        """
+        if hasattr(self.engine, "shutdown"):
+            logger.info("Shutting down vLLM engine")
+            self.engine.shutdown()
+
+
+class vLLMEngineStageUDF(StatefulStageUDF):
+    def __init__(
+        self,
+        data_column: str,
+        model: str,
+        engine_kwargs: Dict[str, Any],
+        task_type: str = "generate",
+        runtime_env: Optional[Dict[str, Any]] = None,
+        max_pending_requests: Optional[int] = None,
+    ):
+        """
+        Initialize the vLLMEngineStageUDF.
+
+        Args:
+            data_column: The data column name.
+            engine_kwargs: The kwargs to pass to the vLLM engine.
+            task_type: The task to use for the vLLM engine (e.g., "generate", "embed", etc).
+            runtime_env: The runtime environment to use for the vLLM engine.
+            max_pending_requests: The maximum number of pending requests. If None,
+                it will be set to 1.1 * max_num_seqs * pipeline_parallel_size.
+        """
+        super().__init__(data_column)
+        self.model = model
+
+        # Setup runtime env.
+        self.runtime_env = runtime_env or {}
+
+        # Setup vLLM engine kwargs.
+        self.task_type = task_type
+        self.engine_kwargs = self.normalize_engine_kwargs(task_type, engine_kwargs)
+
+        # Set up the max pending requests.
+        pp_size = self.engine_kwargs.get("pipeline_parallel_size", 1)
+        self.max_pending_requests = max_pending_requests or math.ceil(
+            self.engine_kwargs["max_num_seqs"] * pp_size * 1.1
+        )
+        if self.max_pending_requests > 0:
+            logger.info("Max pending requests is set to %d", self.max_pending_requests)
```

Comment:
can we use f-string wherever it's not perf critical ? Most of this library I think falls under the category of non performance critical. Engine is performance critical on the other hand.

Reply:
I believe this is a linting rule (log-no-lazy). Using f-string in logger will violate lots of linting tools, and it's hard to guarantee no f-string is used by performance critical logging if we rely on contributors and reviewers to judge. The easiest and the most reliable way is enforcing everything.